### PR TITLE
OCPBUGS-18933: Fix reint-on-failed issues

### PR DESCRIPTION
**Pull request by Vincent056**
This changes prevent all node reinit when reint-on-failed is being used.

[OCPBUGS-18933](https://issues.redhat.com/browse/OCPBUGS-18933)

### CODE REVIEW

**Comment by Vincent056** [comment]
failed to get cluster,
/retest

**Comment by BhargaviGudi** [comment]
Verification passed on 4.14.0-0.nightly-2023-09-26-124507 + file-integrity-operator from PR #436 

1. Install FIO
2. Create fileintegrity
3. Trigger fileintegrity failures on two nodes
```
$ oc get fileintegritynodestatuses.fileintegrity.openshift.io 
NAME                                                              NODE                                        STATUS
example-fileintegrity-ip-10-0-27-104.us-east-2.compute.internal   ip-10-0-27-104.us-east-2.compute.internal   Succeeded
example-fileintegrity-ip-10-0-44-20.us-east-2.compute.internal    ip-10-0-44-20.us-east-2.compute.internal    Failed
example-fileintegrity-ip-10-0-54-131.us-east-2.compute.internal   ip-10-0-54-131.us-east-2.compute.internal   Failed
example-fileintegrity-ip-10-0-80-201.us-east-2.compute.internal   ip-10-0-80-201.us-east-2.compute.internal   Succeeded
example-fileintegrity-ip-10-0-83-23.us-east-2.compute.internal    ip-10-0-83-23.us-east-2.compute.internal    Succeeded
example-fileintegrity-ip-10-0-9-203.us-east-2.compute.internal    ip-10-0-9-203.us-east-2.compute.internal    Succeeded
```
4. oc annotate fileintegrities/<fileintegrity-name> file-integrity.openshift.io/re-init-on-failed=
```
$ oc annotate fileintegrity example-fileintegrity file-integrity.openshift.io/re-init-on-failed=
fileintegrity.fileintegrity.openshift.io/example-fileintegrity annotated
```
fileintegritynodestatus for failed nodes are marked are Succeeded and db is re-inted only on failed nodes
```
$ oc get fileintegritynodestatuses.fileintegrity.openshift.io 
NAME                                                              NODE                                        STATUS
example-fileintegrity-ip-10-0-27-104.us-east-2.compute.internal   ip-10-0-27-104.us-east-2.compute.internal   Succeeded
example-fileintegrity-ip-10-0-44-20.us-east-2.compute.internal    ip-10-0-44-20.us-east-2.compute.internal    Succeeded
example-fileintegrity-ip-10-0-54-131.us-east-2.compute.internal   ip-10-0-54-131.us-east-2.compute.internal   Succeeded
example-fileintegrity-ip-10-0-80-201.us-east-2.compute.internal   ip-10-0-80-201.us-east-2.compute.internal   Succeeded
example-fileintegrity-ip-10-0-83-23.us-east-2.compute.internal    ip-10-0-83-23.us-east-2.compute.internal    Succeeded
example-fileintegrity-ip-10-0-9-203.us-east-2.compute.internal    ip-10-0-9-203.us-east-2.compute.internal    Succeeded
$ for node in `oc get node --no-headers | awk '{print $1}'`; do oc debug node/$node -- chroot /host ls -ltr /etc/kubernetes/; done
Starting pod/ip-10-0-27-104us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 3640
-rw-------.  1 root root    9249 Sep 29 05:17 kubeconfig
drwxr-xr-x.  3 root root      19 Sep 29 05:18 cni
drwxr-xr-x.  3 root root      20 Sep 29 05:18 kubelet-plugins
-rw-r--r--.  1 root root    8332 Sep 29 05:28 kubelet-ca.crt
drwxr-xr-x. 16 root root    4096 Sep 29 05:34 static-pod-resources
drwxr-xr-x.  2 root root     129 Sep 29 05:34 manifests
-rw-r--r--.  1 root root     107 Sep 29 05:42 apiserver-url.env
-rw-r--r--.  1 root root    1123 Sep 29 05:42 ca.crt
-rw-r--r--.  1 root root       0 Sep 29 05:42 cloud.conf
-rw-r--r--.  1 root root    2906 Sep 29 05:42 kubelet.conf
-rw-------.  1 root root 1835684 Sep 29 07:20 aide.db.gz.new
-rw-------.  1 root root 1835684 Sep 29 07:20 aide.db.gz
-rw-------.  1 root root     777 Sep 29 08:18 aide.log.new
-rw-------.  1 root root     777 Sep 29 08:18 aide.log

Removing debug pod ...
Starting pod/ip-10-0-44-20us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 7224
-rw-------. 1 root root    6050 Sep 29 05:22 kubeconfig
drwxr-xr-x. 3 root root      19 Sep 29 05:23 cni
drwxr-xr-x. 3 root root      20 Sep 29 05:23 kubelet-plugins
drwxr-xr-x. 2 root root       6 Sep 29 05:23 manifests
drwxr-xr-x. 3 root root      24 Sep 29 05:24 static-pod-resources
-rw-r--r--. 1 root root    8332 Sep 29 05:28 kubelet-ca.crt
-rw-r--r--. 1 root root    1123 Sep 29 05:42 ca.crt
-rw-r--r--. 1 root root       0 Sep 29 05:42 cloud.conf
-rw-r--r--. 1 root root    2906 Sep 29 05:42 kubelet.conf
-rw-------. 1 root root    1004 Sep 29 07:28 aide.log.backup-20230929T07_28_05
-rw-------. 1 root root 1835388 Sep 29 07:28 aide.db.gz.backup-20230929T07_28_05
-rw-------. 1 root root 1835409 Sep 29 08:16 aide.db.gz.backup-20230929T08_16_44
-rw-------. 1 root root    1010 Sep 29 08:16 aide.log.backup-20230929T08_16_44
-rw-------. 1 root root 1835494 Sep 29 08:17 aide.db.gz.new
-rw-------. 1 root root 1835494 Sep 29 08:17 aide.db.gz
-rw-------. 1 root root     777 Sep 29 08:17 aide.log
-rw-------. 1 root root       0 Sep 29 08:17 aide.log.new

Removing debug pod ...
Starting pod/ip-10-0-54-131us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 7236
-rw-------.  1 root root    9249 Sep 29 05:16 kubeconfig
drwxr-xr-x.  3 root root      19 Sep 29 05:17 cni
drwxr-xr-x.  3 root root      20 Sep 29 05:17 kubelet-plugins
-rw-r--r--.  1 root root    8332 Sep 29 05:28 kubelet-ca.crt
drwxr-xr-x. 17 root root    4096 Sep 29 05:35 static-pod-resources
drwxr-xr-x.  2 root root     129 Sep 29 05:35 manifests
-rw-r--r--.  1 root root     107 Sep 29 05:42 apiserver-url.env
-rw-r--r--.  1 root root    1123 Sep 29 05:42 ca.crt
-rw-r--r--.  1 root root       0 Sep 29 05:42 cloud.conf
-rw-r--r--.  1 root root    2906 Sep 29 05:42 kubelet.conf
-rw-------.  1 root root 1835801 Sep 29 07:27 aide.db.gz.backup-20230929T07_27_45
-rw-------.  1 root root    1004 Sep 29 07:27 aide.log.backup-20230929T07_27_45
-rw-------.  1 root root 1835817 Sep 29 08:16 aide.db.gz.backup-20230929T08_16_53
-rw-------.  1 root root    1010 Sep 29 08:16 aide.log.backup-20230929T08_16_53
-rw-------.  1 root root 1835895 Sep 29 08:17 aide.db.gz.new
-rw-------.  1 root root 1835895 Sep 29 08:17 aide.db.gz
-rw-------.  1 root root     777 Sep 29 08:17 aide.log
-rw-------.  1 root root       0 Sep 29 08:18 aide.log.new

Removing debug pod ...
Starting pod/ip-10-0-80-201us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 3636
-rw-------.  1 root root    9249 Sep 29 05:15 kubeconfig
drwxr-xr-x.  3 root root      19 Sep 29 05:16 cni
drwxr-xr-x.  3 root root      20 Sep 29 05:16 kubelet-plugins
-rw-r--r--.  1 root root    8332 Sep 29 05:28 kubelet-ca.crt
drwxr-xr-x. 16 root root    4096 Sep 29 05:37 static-pod-resources
drwxr-xr-x.  2 root root     129 Sep 29 05:37 manifests
-rw-r--r--.  1 root root     107 Sep 29 05:43 apiserver-url.env
-rw-r--r--.  1 root root    1123 Sep 29 05:43 ca.crt
-rw-r--r--.  1 root root    2906 Sep 29 05:43 kubelet.conf
-rw-r--r--.  1 root root       0 Sep 29 05:43 cloud.conf
-rw-------.  1 root root 1835736 Sep 29 07:20 aide.db.gz.new
-rw-------.  1 root root 1835736 Sep 29 07:20 aide.db.gz
-rw-------.  1 root root     777 Sep 29 08:18 aide.log
-rw-------.  1 root root       0 Sep 29 08:18 aide.log.new

Removing debug pod ...
Starting pod/ip-10-0-83-23us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 3628
-rw-------. 1 root root    6050 Sep 29 05:25 kubeconfig
drwxr-xr-x. 3 root root      19 Sep 29 05:25 cni
drwxr-xr-x. 3 root root      20 Sep 29 05:26 kubelet-plugins
drwxr-xr-x. 2 root root       6 Sep 29 05:26 manifests
drwxr-xr-x. 3 root root      24 Sep 29 05:28 static-pod-resources
-rw-r--r--. 1 root root    8332 Sep 29 05:28 kubelet-ca.crt
-rw-r--r--. 1 root root    1123 Sep 29 05:43 ca.crt
-rw-r--r--. 1 root root    2906 Sep 29 05:43 kubelet.conf
-rw-r--r--. 1 root root       0 Sep 29 05:43 cloud.conf
-rw-------. 1 root root      67 Sep 29 07:19 aide.log.backup-20230929T07_19_55
-rw-------. 1 root root 1835396 Sep 29 07:20 aide.db.gz.new
-rw-------. 1 root root 1835396 Sep 29 07:20 aide.db.gz
-rw-------. 1 root root     777 Sep 29 08:17 aide.log
-rw-------. 1 root root       0 Sep 29 08:18 aide.log.new

Removing debug pod ...
Starting pod/ip-10-0-9-203us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
total 3628
-rw-------. 1 root root    6050 Sep 29 05:22 kubeconfig
drwxr-xr-x. 3 root root      19 Sep 29 05:23 cni
drwxr-xr-x. 3 root root      20 Sep 29 05:23 kubelet-plugins
drwxr-xr-x. 2 root root       6 Sep 29 05:23 manifests
drwxr-xr-x. 3 root root      24 Sep 29 05:25 static-pod-resources
-rw-r--r--. 1 root root    8332 Sep 29 05:28 kubelet-ca.crt
-rw-r--r--. 1 root root    1123 Sep 29 05:42 ca.crt
-rw-r--r--. 1 root root       0 Sep 29 05:42 cloud.conf
-rw-r--r--. 1 root root    2906 Sep 29 05:42 kubelet.conf
-rw-------. 1 root root 1835434 Sep 29 07:20 aide.db.gz.new
-rw-------. 1 root root 1835434 Sep 29 07:20 aide.db.gz
-rw-------. 1 root root     777 Sep 29 08:18 aide.log.new
-rw-------. 1 root root     777 Sep 29 08:18 aide.log

Removing debug pod ...
```

File: pkg/common/util.go [hunk 1]
```diff
@@ -219,14 +219,18 @@ func GetRemovedNodeReinitAnnotation(fi *v1alpha1.FileIntegrity, nodeName string)
...
+			} else {
+				newNodeList := []string{}
+				for _, node := range nodeList {
+					if node == nodeName {
+						continue
+					}
+					newNodeList = append(newNodeList, node)
 				}
-				newNodeList = append(newNodeList, node)
+				ficopy.Annotations[AideDatabaseReinitAnnotationKey] = strings.Join(newNodeList, ",")
```
Comment (rhmdnd) [inline pkg/common/util.go:232]: Do we have a test case that already covers this?

Comment (Vincent056) [reply pkg/common/util.go:232]: I think we need to add that specific test case

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -325,6 +325,109 @@ func TestFileIntegrityPruneBackup(t *testing.T) {
...
+			}
+
+			if !containsUncompressedScanFailLog(data) {
+				t.Errorf("configMap '%s' does not have a failure log. Got: %#v", result.ResultConfigMapName, data)
+			}
+		}
+	}
+
+	expectedMetrics = map[string]int{}
+	// Each node should have a metric set at 1 for node_failed
```
Comment (rhmdnd) [inline tests/e2e/e2e_test.go:401]: Does it make sense to check that the master nodes don't have this metric?

Comment (Vincent056) [reply tests/e2e/e2e_test.go:401]: let's address that in a sperate PR

**Pull request merged**